mantle/kola/tests/misc/network.go

```diff
@@ -172,3 +328,145 @@ func NetworkInitramfsSecondBoot(c cluster.TestCluster) {
+		return fmt.Errorf("interface mismatch: %s MAC Address interface %s and IP Address %s interface %s should be the same", setMac, macInterface, expectedIP, ipInterface)
+	}
+
+	return nil
+}
+
+func getInterfaces(c cluster.TestCluster, machineIdx int) ([]string, error) {
+	interfaceList := []string{}
+	m := c.Machines()[machineIdx]
+
+	interfacesOutput := string(c.MustSSH(m, "ls /sys/class/net"))
+	interfaceList = strings.Split(interfacesOutput, "\n")
+
+	// a total of four NICS: lo, default interface, and two secondary NICs.
+	expectedInterfacesNum := 4
+	if len(interfaceList) != expectedInterfacesNum {
+		return []string{}, fmt.Errorf("multiple interfaces not created on machine: %v", interfaceList)
+	}
+	return interfaceList, nil
+}
+
+func getInterfaceByMac(c cluster.TestCluster, machineIdx int, interfacesList []string) (map[string]string, error) {
+	ifaceMACAddressMap := map[string]string{}
+	m := c.Machines()[machineIdx]
+
+	for _, iface := range interfacesList {
+		if iface == "lo" {
+			continue
+		}
+		macAddress := string(c.MustSSH(m, fmt.Sprintf("cat /sys/class/net/%s/address", iface)))
+		ifaceMACAddressMap[macAddress] = iface
+	}
+	return ifaceMACAddressMap, nil
+}
+
+func getInterfaceByIpAddress(c cluster.TestCluster, machineIdx int, interfacesList []string) (map[string]string, error) {
+	ifaceMACAddressMap := map[string]string{}
+	m := c.Machines()[machineIdx]
+
+	for _, iface := range interfacesList {
+		if iface == "lo" {
+			continue
+		}
+		ipAddress := string(c.MustSSH(m, fmt.Sprintf("ip addr show %s | grep inet | grep -v inet6 |awk '{print $2}'|awk '{split($0,a,\"/\"); print a[1]}'", iface)))
+		ifaceMACAddressMap[ipAddress] = iface
+	}
+	return ifaceMACAddressMap, nil
+}
+
+func addKernelArgs(c cluster.TestCluster, m platform.Machine, args []string) {
+	if len(args) == 0 {
+		return
+	}
+
+	rpmOstreeCommand := "sudo rpm-ostree kargs"
+	for _, arg := range args {
+		rpmOstreeCommand = fmt.Sprintf("%s --append %s", rpmOstreeCommand, arg)
+	}
+
+	c.MustSSH(m, rpmOstreeCommand)
+
+	err := m.Reboot()
+	if err != nil {
+		c.Fatalf("failed to reboot the machine: %v", err)
+	}
+}
+
+func setupMultipleNetworkTest(c cluster.TestCluster, primaryMac, secondaryMac string) {
+	var m platform.Machine
+	var err error
+
+	options := platform.QemuMachineOptions{
+		SecondaryNics: 2,
+	}
+
+	switch pc := c.Cluster.(type) {
+	// These cases have to be separated because when put together to the same case statement
+	// the golang compiler no longer checks that the individual types in the case have the
+	// NewMachineWithQemuOptions function, but rather whether platform.Cluster
+	// does which fails
+	case *unprivqemu.Cluster:
+		m, err = pc.NewMachineWithQemuOptions(nil, options)
+	default:
+		panic("unreachable")
+	}
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	// Add karg needed for the ignition to configure the network properly.
+	addKernelArgs(c, m, []string{fmt.Sprintf("macAddressList=%s,%s", primaryMac, secondaryMac)})
```

Comment:
the mac addresses is something that is given as an input to the image.
We want to be as close to the real life case, in order to find out issues as soon as possible.

Reply:
Ack thanks. Reading https://github.com/RHsyseng/rhcos-slb now it makes more sense. I think there's ways we could improve the workflow here. E.g. using `coreos-installer install --append-karg` to pass that info from the PXE install to the installed system. But right now this would require overriding the default coreos-installer service to customize it.  `/boot` should be considered owned by the OS.

Re. the karg name, maybe it should be something more specific, like `ovs.mac_addresses` or so?

Another approach BTW which short-circuits a lot of this is to just unify all the MAC addresses in the same table and shove that in the Ignition config (e.g. as a day-1 MC), and the OVS script then can just look up in the table to know which MAC addresses to use for the node it's running on. Then e.g. the PXE config is no longer host-specific, there's no need to write to `/boot`, users just need to maintain one MC, which is tracked in etcd, etc...

Anyway, this is all feedback about https://github.com/RHsyseng/rhcos-slb I guess and not this code here that's trying to simulate it. :)

Reply:
Sure, sounds good!